File: 2025/R/solve_day2_part2.R
solve_day2_part2 <- function(input) {
  # back to one line
  ranges <- paste0(input, collapse = "")
  ranges <- strsplit(ranges, ",", fixed = TRUE) |>
    unlist()

  ranges <- strsplit(ranges, "-", fixed = TRUE)

  invalid_total <- 0

  for (range in ranges) {
    ids <- range[[1]]:range[[2]]

    invalid_ids <- vapply(ids, is_invalid, logical(1))

    invalid_total <- invalid_total + sum(ids[invalid_ids])
  }

  invalid_total
}

is_invalid <- function(id) {
  nchars <- nchar(id)

  for (i in seq_len(nchars - 1)) {
    for (j in i:nchars) {
      substr_len <- j - i + 1

      if ((nchars - substr_len) < substr_len) {
        break
      }

      if (nchars %% substr_len != 0) {
        next
      }

      substr_chars <- substr(id, i, j)
      check_count <- (nchars - substr_len) / substr_len

      checks <- vapply(
        seq_len(check_count),
        \(multi) {
          step <- multi * substr_len
          substr(id, i + step, j + step) == substr_chars
        },
        logical(1)
      )

      if (all(checks)) return(TRUE)
    }
  }

  FALSE
}

File: 2025/python/solve_day2_part2.py
def solve_day2_part2(text):
  ranges = text[0].split(",")
  ranges = [str.split("-") for str in ranges]

  invalid_total = 0

  for nums in ranges:
    for num in range(int(nums[0]), int(nums[1])+1):
      if invalid(num):
        invalid_total += num

  return(invalid_total)

def invalid(num):
    num = str(num)
    nchars = len(num)

    for str_len in range(1, len(num)//2 + 1):
        invalid = True
        if (nchars - str_len) < 0 :
            break

        if (nchars % str_len) != 0:
            continue

        substr_chars = num[:str_len]
        count_to_check = (nchars - str_len) // str_len + 1

        for multi in range(1, count_to_check):
            step = multi * str_len
            if num[step:step+str_len] != substr_chars:
                invalid = False
                break

        if invalid:
            return(True)

    return(False)
